### PR TITLE
berryjen: binary search tree exercise

### PR DESCRIPTION
**Pull request by berryjen**
This exercise started with an intro to understanding `npm` & `scripts`. From there, the exercise began with trying to sort out the `searchRecursive` function within `class BinarySearchTree`. Breaking down the problem into pseudo-code went fine, but it was the next step of translating that pseudo-code into actual code that was more challenging. A TDD approach was integrated into this exercise, and while running` ./tests`, the test file didn't run. Mini-lesson on absolute vs relative path files and folder naming conventions have been really useful to navigate file locations in the terminal running commands like `pwd` & `ls`. Remembering to include node at the beginning of each command before running the local file was also a nuanced habit. Next, placement of console.log statements & params passed into Class constructors were also focused on. I haven't worked much with `Classes` & `constructors`, so its behaviors are new to me. A `constructor` can take an `object` with multiple properties as a param or positional arguments. I created a new file `learning.js` to play around with Classes, constructors and instances to explore the relationships among them. 
The next learning moment surfaced when a persistent bug regarding `TypeError: Cannot read property 'val' of undefined `kept recurring. This was due to the fact that I'd forgotten to consider the edge case of when a `node` is `null`. In summary there are 5 possible outcomes: 

- An empty (`null)` node
- an `undefined` node
- The `search` value matches the node value
- The `search `value is greater than the node value
- The `search` value is less than the node value
For the` search` function, we called` researchRecursive` with arguments of `id` & `this.root` were passed to taken into account of a `null` root. 

It was important to point out that I had originally returned the value of `inventoryCount `of the `currentNode` within the `updateInventory` function, when this wasn't required from the sprint. This is another learning moment where I can take into future account to ask for clarifications (instead of going ahead with my own assumptions ) at the beginning of the AGILE cycle. 
At this point, I had managed to get 3/4 tests passing with the remaining 1 test failing - `should update inventory count for the product for given ID`.- I could not debug why that test wasn't passing despite the 3 functions working harmoniously within` class BinarySearchTree`. A helpful hint was nudged my way to look into the test file- particularly if the `const Product` had the same `name` & `ID` that matched. After investigating the issue further, the bug resided there: there was a mismatch between the `const name` & `ID`

### CODE REVIEW

File: binary-search-tree/node-js/inventory-management.js [hunk 1]
```diff
@@ -18,16 +18,37 @@ class BinarySearchTree {
         this.root = null;
     }
 
-    search(id) {
-        console.error("you must implement search");
+    search(id, currentNode) {
```
Comment (doodlesbykumbi) [inline binary-search-tree/node-js/inventory-management.js]: You shouldn't need to change the signature of the `search` method

File: binary-search-tree/node-js/inventory-management.js [hunk 2]
```diff
@@ -18,16 +18,37 @@ class BinarySearchTree {
...
+        if (id === null) {
```
Comment (doodlesbykumbi) [inline binary-search-tree/node-js/inventory-management.js]: I get what you're trying to do here by making sure we only process valid input, but I think you can reasonable always assume that `id` will be a valid value whether or not a node with the same value exists in the binary search tree.

File: binary-search-tree/node-js/inventory-management.js [hunk 3]
```diff
@@ -18,16 +18,37 @@ class BinarySearchTree {
...
+            return null;
+        } else {
+            return this.searchRecursive(id, currentNode)
```
Comment (doodlesbykumbi) [inline binary-search-tree/node-js/inventory-management.js]: Now that you change the signature of the `search` method, which you shouldn't need to, imagine that this method is called by something that only knows the original signature e.g.

```javascript
bst.search(9)
```

Now inside the function `currentNode` will be set to `undefined`. That would lead to all sorts of issues, including the fact that we're no longer starting with `currentNode` as the root node.

File: binary-search-tree/node-js/inventory-management.js [hunk 4]
```diff
@@ -18,16 +18,37 @@ class BinarySearchTree {
...
+        // 3. Determine if we should go left or right
+        // 4. Update the current node based on the decision from (3). Rinse and repeat from (2)
+    searchRecursive(id, currentNode) {
+        if(currentNode.val.id === id ) {
+            return currentNode
+        }
+        else if (id > currentNode.val.id) {
+            return this.searchRecursive(id, currentNode.right)
+        }
+        else (id < currentNode.val.id)
```
Comment (doodlesbykumbi) [inline binary-search-tree/node-js/inventory-management.js]: An `else` can't take a condition so this bit might not do what you think it does. It probably has unintended behavior because the interpreter 

1. Sees `(id < currentNode.val.id)` as the block of the `else`
2. See the `return ...`, immediately under (1), as separate from the else block.

Please clean up the syntax here to avoid misinterpretation

File: binary-search-tree/node-js/inventory-management.js [hunk 5]
```diff
@@ -18,16 +18,37 @@ class BinarySearchTree {
...
+            return this.searchRecursive(id, currentNode)
+        }
     }
-
+        // console.error("you must implement search");
+        // 1. Start from the root node
+        // 2. Compare the current node with the search value. Return early if found
+        // 3. Determine if we should go left or right
+        // 4. Update the current node based on the decision from (3). Rinse and repeat from (2)
+    searchRecursive(id, currentNode) {
```
Comment (doodlesbykumbi) [inline binary-search-tree/node-js/inventory-management.js:25]: I urge you to consider the thread of execution of this method when `currentNode` is a leaf node (has no children) and doesn't have the same `id` as the search value.

Here's such a thread of execution

1. `searchRecursive` is called with `id=9` and `currentNode` is a leaf node with `val.id=5`.
2. The first `if` statement is not true since `currentNode.val.id` is not equal to `id`. So we move to the next `else if` statement.
3. The second `else if` statement is true since `id` is greater than `currentNode.val.id` (9 > 5). So we call `searchRecursive` again with `id=9` and `currentNode=currentNode.right`.
   - This creates a new stack frame on the call stack with a new set of parameters: `id=9` and `currentNode` is `null` because `currentNode.right` is `null`.
   - The function then enters the new stack frame with these new parameters and the thread of execution continues from the beginning of the function with the new set of parameters.
   - Since `currentNode` is `null`, the next conditional statement `if(currentNode.val.id === id )` evaluates to `null.val.id === id` which is trying to access the `val` property of `null`, resulting in the error "TypeError: Cannot read property `val` of null."


Being able to follow the thread of execution is important!

**Review by doodlesbykumbi — COMMENTED** [review]
Left some comments

File: binary-search-tree/node-js/inventory-management.js (hunk 5 above)
Comment (berryjen) [reply binary-search-tree/node-js/inventory-management.js:25]: sounds good, thanks for the feedback, am on it already

File: binary-search-tree/node-js/inventory-management.js [hunk 6]
```diff
@@ -10,6 +10,10 @@ class Product {
     constructor({id, inventoryCount}) {
         this.id = id;
         this.inventoryCount = inventoryCount;
+        this.inventoryCount = 0;
+    }
+    inventoryCount() {
+        this.inventoryCount ++; 
     }
```
Comment (doodlesbykumbi) [inline binary-search-tree/node-js/inventory-management.js]: The exercise spec doesn't ask you to do this. Please reflect on why you added this, and how it might be contributing to the failing test

File: binary-search-tree/node-js/inventory-management.js [hunk 7]
```diff
@@ -19,15 +23,40 @@ class BinarySearchTree {
...
+    };
+    
     updateInventory(id, inventoryCount) {
-        console.error("you must implement updateInventory");
+        const currentNode = this.searchRecursive(id, this.root);
+        if (currentNode === null) {
+            return;
+        }
+        currentNode.inventoryCount = inventoryCount;
+        return currentNode.inventoryCount;
```
Comment (doodlesbykumbi) [inline binary-search-tree/node-js/inventory-management.js]: The exercise spec doesn't specify a return value for `updateInventory`. Worth reflecting on why you chose to return this value. 

I'm not saying this is wrong. It's just good to be intentional and have a good reason backing a particular design decision.

File: binary-search-tree/node-js/inventory-management.js [hunk 8]
```diff
@@ -19,15 +23,40 @@ class BinarySearchTree {
...
+        }
+        else if(currentNode.val.id === id ) {
+            return currentNode.val;
+        }
+        else if (id > currentNode.val.id) {
+            return this.searchRecursive(id, currentNode.right);
+        }
+        else {
+            return this.searchRecursive(id, currentNode.left);
+        }
```
Comment (doodlesbykumbi) [inline binary-search-tree/node-js/inventory-management.js]: Nice! I see that you're getting a better understanding of the exercise. This is good and it works, the `search` method test cases are passing. Here is some feedback to make it better:

There's a potential issue with conditional branches, specifically the `else` branch that could lead to unintended behavior in certain edge cases.

It's important to handle the cases you care about explicitly to avoid unexpected behavior. Currently, your code handles `id < currentNode.val.id` implicitly within the `else` branch. However, if `currentNode` is `undefined` (not `null`), this can lead to an undesired outcome. Specifically you'll get an an infinite loop (leaving it as an exercise for you to confirm).

To address this, I recommend adding an explicit `else if (id < currentNode.val.id)` branch

**Review by doodlesbykumbi — COMMENTED** [review]
Resolved previous comments and left new batch of feedback.

I wanted to acknowledge your efforts on this exercise. I can already see elements of growth and better understanding, I hope you can too. Keep going!

File: binary-search-tree/node-js/inventory-management.js (hunk 6 above)
Comment (berryjen) [reply binary-search-tree/node-js/inventory-management.js]: "The update_inventory method takes a product ID as its argument. The method should find the node with the corresponding id attribute in the binary search tree and update the inventory_count attribute of the Product object stored in the val attribute of that node." In the ReadMe, the task is asking us to update `inventoryCount`

File: binary-search-tree/node-js/inventory-management.js (hunk 7 above)
Comment (berryjen) [reply binary-search-tree/node-js/inventory-management.js]: the task is asking us to update `inventoryCount`...so doesn't that mean we need to return a value for `inventoryCount`? Unless i've misinterpreted the problem..

File: binary-search-tree/node-js/inventory-management.js (hunk 6 above)
Comment (doodlesbykumbi) [reply binary-search-tree/node-js/inventory-management.js]: I think this can be a lesson
1. In situational awareness. If you look around the codebase and observe how `update_inventory` is being called you'll notice that it takes two arguments, product ID and new inventory count.
2. In asking for clarification from the spec

File: binary-search-tree/node-js/inventory-management.js (hunk 7 above)
Comment (doodlesbykumbi) [reply binary-search-tree/node-js/inventory-management.js]: No, it doesn't necessarily mean it needs to be return. The task is purely about updating the `inventoryCount`.

Though, if we're talking about being intentional, a good reason for returning the updated inventory count on success and `undefined` otherwise, as your function is currently doing, would be to make it clear to the caller that the operation was successful. So that if I call `bst.updateInventory(6, 20)` my intention is to update the inventory count of the `Product` with `id=6` to `20`. If the product is found and its inventory count is updated then the return value is `20`, otherwise the return value is `undefined`. Without an explicit return value the function would otherwise always return `undefined` and I wouldn't be able to immediately make heads or tales of the outcome.

File: binary-search-tree/node-js/inventory-management.js (hunk 8 above)
Comment (berryjen) [reply binary-search-tree/node-js/inventory-management.js]: ahh... ok. Originally i did have `else if (id < currentNode.val.id)` and then else { return;}

File: binary-search-tree/node-js/inventory-management.js (hunk 6 above)
Comment (berryjen) [reply binary-search-tree/node-js/inventory-management.js]: yes because i knew that it takes `inventoryCount` as a param, therefore in my head it made sense that a return value for it

File: binary-search-tree/node-js/inventory-management.js (hunk 7 above)
Comment (berryjen) [reply binary-search-tree/node-js/inventory-management.js]: i am explicitly returning value of `inventoryCount`....i'm confused with your explanation....

File: binary-search-tree/node-js/inventory-management.js [hunk 9]
```diff
@@ -10,6 +10,8 @@ class Product {
     constructor({id, inventoryCount}) {
         this.id = id;
         this.inventoryCount = inventoryCount;
+        this.inventoryCount ++;
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/inventory-management.js]: It turns out the last test is passing because of a fluke. I explain below.

If you look at the test case for `updateInventory`, the idea is that it 
1. Creates a  `Product` which starts with an `inventoryCount` of `7`
2. Updates the inventory count by making the call `bst.updateInventory(6, 8)`
3. Asserts that the updated inventory count is `8`

It just so happens that line 13 increments the `inventoryCount`, and that's why the test passes. If we instead change the update value from `8` to `27` the test will fail.

All that to say, line 13 is not relevant to the solution and should be removed. Remove line 13 and look at the test failure to understand why it is failing.

**Review by asideofcode-dev — COMMENTED** [review]
It turns out the last test is passing because of a fluke. I explain in a comment.

File: binary-search-tree/node-js/inventory-management.js (hunk 9 above)
Comment (berryjen) [reply binary-search-tree/node-js/inventory-management.js]: i indeed took a look at the previous failing tests, and noticed that 7 !== 8, so logically i thought i was off by 1 count, hence why i implemented line 13

Comment (asideofcode-dev) [reply binary-search-tree/node-js/inventory-management.js]: It's good that you looked at the failing test, just the approach to addressing it needs some work. 

The `updateInventory` method looks sound. Did you check to see if there's any issues with the test itself? For instance, does product six actually have `id=6` 🙈

Comment (berryjen) [reply binary-search-tree/node-js/inventory-management.js]: i did, and lines 28-47 on `test.js` look rather odd....productSix has 4 other sub-products under it...?!
 `const productSix = new Product({ id: 14, inventoryCount: 7 });` this doesn't match

File: binary-search-tree/node-js/inventory-management.js [hunk 10]
```diff
@@ -10,6 +10,8 @@ class Product {
     constructor({id, inventoryCount}) {
         this.id = id;
         this.inventoryCount = inventoryCount;
+        
+
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/inventory-management.js]: cleanup

File: binary-search-tree/node-js/inventory-management.js [hunk 11]
```diff
@@ -19,15 +21,45 @@ class BinarySearchTree {
     }
 
     search(id) {
-        console.error("you must implement search");
-    }
-
+            return this.searchRecursive(id, this.root);
+        }
+        // console.error("you must implement search");
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/inventory-management.js]: cleanup

File: binary-search-tree/node-js/inventory-management.js [hunk 12]
```diff
@@ -19,15 +21,45 @@ class BinarySearchTree {
...
+        const currentNode = this.searchRecursive(id, this.root);
+        if (currentNode === null) {
+            return;
+        }
+        currentNode.inventoryCount = inventoryCount;
     }
 
     buildTree(products) {
-        console.error("you must implement buildTree");
+        // console.error("you must implement buildTree");
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/inventory-management.js]: cleanup

File: binary-search-tree/node-js/inventory-management.js [hunk 13]
```diff
@@ -76,7 +108,7 @@ class BinarySearchTree {
         }
     }
 }
-
+// console.log(updateInventory(id,inventoryCount));
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/inventory-management.js]: cleanup

File: binary-search-tree/node-js/learning.js [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+        this.count ++; 
+    }
+}
+const lotion = new Product({id:6, inventoryCount:3, name:'lotion'});
+lotion.inventoryCount = 4
+console.log(lotion.id);
+console.log(lotion.inventoryCount);
+
+let currentNode = null;
+console.log(currentNode.val.id);
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/learning.js]: cleanup

File: binary-search-tree/node-js/package.json [hunk 15]
```diff
@@ -4,8 +4,9 @@
   "description": "",
   "main": "inventory-management.js",
   "scripts": {
-    "test": "mocha inventory-management.test.js"
-  },  
+    "test": "mocha inventory-management.test.js", 
+    "gem":  "echo this is gem"
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/package.json]: cleanup

File: binary-search-tree/node-js/test [hunk 16]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 
-npm run test
+ npm run test
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/test:2]: cleanup

**Review by asideofcode-dev — COMMENTED** [review]
Nice work @berryjen, looks like all the tests are passing 🎉. Stay persevering!

Some additional tasks for you before we close out the PR
1. It's generally a good idea to ensure that your pull request only contains the changes that address the goals of the pull request. 

   While working through the PR you might make some modifications that were useful in order for you to figure things out but that you wouldn't want to get get merged e.g. additional whitespace, unnecessary comments and maybe some code that you used for learning. It's a good idea as you prepare to close out the PR to clean up the codebase. 
   
   I've left some comments with just the word `cleanup` to point to places where you might want to carry out the clean up.
2. There's been a bit of back and forth for us to get here. I think it would be good for you to take time to write a summary that captures the important parts of how we got here, and some takeaways you might have from the exercise. You can add this to the PR description (you can edit the PR description by clicking on the 3 dots to the right of it, and clicking edit)

File: binary-search-tree/node-js/inventory-management.js [hunk 17]
```diff
@@ -9,7 +9,7 @@ class TreeNode {
 class Product {
     constructor({id, inventoryCount}) {
         this.id = id;
-        this.inventoryCount = inventoryCount;
+        this.inventoryCount = inventoryCount;
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/inventory-management.js]: Remove the newly introduced additional space

File: binary-search-tree/node-js/inventory-management.js [hunk 18]
```diff
@@ -19,15 +19,40 @@ class BinarySearchTree {
     }
 
     search(id) {
-        console.error("you must implement search");
-    }
-
+            return this.searchRecursive(id, this.root);
+        }
```
Comment (asideofcode-dev) [inline binary-search-tree/node-js/inventory-management.js]: The indentation is off here

File: binary-search-tree/node-js/test (hunk 16 above)
Comment (asideofcode-dev) [inline binary-search-tree/node-js/test:2]: Remove the newly introduced additional space

**Review by asideofcode-dev — COMMENTED** [review]
Left a few more comments re: cleaning up.

File: binary-search-tree/node-js/test (hunk 16 above)
Comment (berryjen) [reply binary-search-tree/node-js/test:2]: finished the learning summary (see above) and removed extra spaces for clean code maintenance :) Great exercise, learned heaps!

**Review by asideofcode-dev — COMMENTED** [review]
Great work, loved the perseverance  @berryjen  ✅ 🎉 !!! It's the consistent effort that gets you there.

I'm going to close out the PR now.

I'd like to present some bonus-round tasks for the exercise:

- [ ] Try refactoring the `search` method to use iteration vs recursion
- [ ] Try implementing your own first of `buildTree` and, think about and look into the pros and cons of different approaches to building the tree. "balanced binary search tree" is a term that might pop up. Happy researching!

If you'd like your attempt to be reviewed feel free to create another PR
